main/ControllerTask.cpp

```diff
@@ -23,99 +24,174 @@
 
 #include "Fuzzyficator.hpp"
 #include "TakagiTsugenoController.hpp"
+#include "MamdaniController.hpp"
 #include "pid.hpp"
 #include "pid.cpp"
+#include "Filters.hpp"
 #include "k_values.h"
 #include "pwm.h"
 
-float bezierCurve(float t, float P0, float P1, float P2, float P3) {
+using namespace std::chrono;
+
+inline float bezierCurve(float t, float P0, float P1, float P2, float P3) {
+	const float oneMinusT = 1-t;
+	const float oneMinusTSquared = oneMinusT*oneMinusT;
+	const float oneMinusTCubed   = oneMinusTSquared*oneMinusT;
+	const float tSquared  = t*t;
+	const float tCubed    = tSquared*t;
 	return
-		  P0*(1-t)*(1-t)*(1-t) +
-		3*P1*(1-t)*(1-t)*t +
-		3*P2*(1-t)*t*t +
-		  P3*t*t*t;
+		  P0*oneMinusTCubed +
+		3*P1*oneMinusTSquared*t +
+		3*P2*oneMinusT*tSquared +
+		  P3*tCubed;
 }
 
 class ControllerTask : public Task {
 private:
 	const QueueHandle_t refer_speed_q;
 	const QueueHandle_t motor_speed_q;
+	const ledc_channel_t pwm_channel;
+
+	const QueueHandle_t _tel_ref_speed_q;
+	const QueueHandle_t _tel_motor_speed_q;
+	const QueueHandle_t _tel_error_q;
+	const QueueHandle_t _tel_error_der_q;
+	const QueueHandle_t _tel_control_signal_q;
+	const QueueHandle_t _tel_exec_time_q;
+
+	static const TickType_t QUEUE_TIMEOUT = 10;
+
+	// taskFunction constants:
+	static constexpr float OUT_MIN = 0.17f;
+	static constexpr float OUT_MAX = 0.95f;
+	static constexpr float U_MIN = 0.0f;
+	static constexpr float U_MAX = 1.0f;
+
+	static constexpr float CRR_CERO     = 0.00;
+	static constexpr float CRR_PEQUE    = 0.25;
+	static constexpr float CRR_MEDIA    = 0.50;
+	static constexpr float CRR_MODERADA = 0.75;
+	static constexpr float CRR_GRANDE   = 1.00;
+
+	static constexpr float K_I = 0.02f;
+	static constexpr float U_RC = 0.80;
+	static constexpr float U_ALPHA = SAMPLE_TIME_s / (SAMPLE_TIME_s + U_RC);
 
 public:
 	ControllerTask(
 		const char *name,
 		uint32_t    stack_size,
 		QueueHandle_t _refer_speed_q,
-		QueueHandle_t _motor_speed_q
+		QueueHandle_t _motor_speed_q,
+		ledc_channel_t _pwm_channel,
+		QueueHandle_t tel_ref_speed_q,
+		QueueHandle_t tel_motor_speed_q,
+		QueueHandle_t tel_error_q,
+		QueueHandle_t tel_error_der_q,
+		QueueHandle_t tel_control_signal_q,
+		QueueHandle_t tel_exec_time_q
 	)
 	: Task(name, stack_size, 2),
-		refer_speed_q(_refer_speed_q), motor_speed_q(_motor_speed_q)
+		refer_speed_q(_refer_speed_q), motor_speed_q(_motor_speed_q),
+		pwm_channel(_pwm_channel),
+		_tel_ref_speed_q(tel_ref_speed_q),
+		_tel_motor_speed_q(tel_motor_speed_q),
+		_tel_error_q(tel_error_q),
+		_tel_error_der_q(tel_error_der_q),
+		_tel_control_signal_q(tel_control_signal_q),
+		_tel_exec_time_q(tel_exec_time_q)
 	{}
 
 	void taskFunction() {
```

Comment:
annotate this function with 'override'. [modernize-use-override,-warnings-as-errors]